### PR TITLE
Redesigned search engine to be a design project

### PR DESCRIPTION
**Pull request by thecardkid**

### CODE REVIEW

File: day07/README.md [hunk 1]
```diff
@@ -1,20 +1,45 @@
...
-
-In `Index.java`:
-
-- Fill in `indexPage(...)`
+# Wikipedia Search Engine Project
+
+## Learning Goals
+In this project, you will learn to:
+* Use persistent storage in the form of Redis
+* Make meaningful design decisions (and be able to defend them!) regarding the tools, given specifications/requirements
```
Comment (davidabrahams) [inline day07/README.md]: "regarding the tools" is a bit vague. I would say "make design decisions regarding which data structures to use given specific problem specifications"

File: day07/README.md [hunk 2]
```diff
@@ -1,20 +1,45 @@
...
+* Design own classes and their appropriate methods
```
Comment (davidabrahams) [inline day07/README.md]: take out "own"

File: day07/README.md [hunk 3]
```diff
@@ -1,20 +1,45 @@
...
+* Learn and understand the functionality of the basic components of a search engine
+* Write testable code, and test effectively
```
Comment (davidabrahams) [inline day07/README.md]: This might be a *bit* much for the project

File: day07/README.md [hunk 4]
```diff
@@ -1,20 +1,45 @@
...
+
+## Preparation
+* Read about search engine [insert resources here]
```
Comment (davidabrahams) [inline day07/README.md]: Address these

File: day07/README.md [hunk 5]
```diff
@@ -1,20 +1,45 @@
...
+* Your Crawler's job is to, given a page, locate and retrieve hyperlinks to other pages. It can keep doing this for as long as you want (you may choose to only look at that page, or all pages 1 hyperlinks away, or all pages 3 hyperlinks away). You decide based on how much time and space is required.
+* A test class that verifies your Crawler's functionality.
+
+### Indexer Requirements
+* Saves and retrieves data to and from your Redis instance.
+  * When saving, it saves a map of search terms to URLs containing those search terms. That is not all it's allowed to save, but that is the minimal.
+  * When retrieving, it must return the top 3 (or fewer) URLs containing the given keyword(s).
+* Given a page, it scans the page for keywords, and add that page's URL to the set of URLs associated with each keyword on Redis.
+* Processes a page in **O(W)**, where W is the number of words on that page. What is the best O(W) you can implement?
+* Queries for pages containg keywords in **O(N)**, where N is the number of relevant URLs.
```
Comment (davidabrahams) [inline day07/README.md]: What's a Query?

File: day07/README.md [hunk 6]
```diff
@@ -1,20 +1,45 @@
...
+* Saves the indexing data of a page to Redis in **O(T)**, where T is the number of unique terms on that page.
+* A test class that verifies your Indexer's functionality.
+
+
+### SearchTool Requirements
+* Supports union, intersection, and difference operators on search terms.
```
Comment (davidabrahams) [inline day07/README.md]: This also could be something we cut/make a stretch goal. Perhaps make `union` required, the others a stretch? I just wanna make sure we don't overload them.

File: day07/README.md (hunk 2 above)
Comment (thecardkid) [inline day07/README.md]: Fixed

File: day07/README.md [hunk 7]
```diff
@@ -1,20 +1,45 @@
...
+* *SearchTool*: takes search terms and boolean operators and returns the most relevant results.
+* *Persistent Storage*: Redis
+
+### Crawler Requirements
+* Your Crawler's job is to, given a page, locate and retrieve hyperlinks to other pages. It can keep doing this for as long as you want (you may choose to only look at that page, or all pages 1 hyperlinks away, or all pages 3 hyperlinks away). You decide based on how much time and space is required.
+* A test class that verifies your Crawler's functionality.
+
+### Indexer Requirements
+* Saves and retrieves data to and from your Redis instance.
+  * When saving, it saves a map of search terms to URLs containing those search terms. That is not all it's allowed to save, but that is the minimal.
```
Comment (thecardkid) [inline day07/README.md]: Fixed

File: day07/README.md [hunk 8]
```diff
@@ -1,20 +1,45 @@
...
+  * When retrieving, it must return the top 3 (or fewer) URLs containing the given keyword(s).
+* Given a page, it scans the page for keywords, and add that page's URL to the set of URLs associated with each keyword on Redis.
+* Processes a page in **O(W)**, where W is the number of words on that page. What is the best O(W) you can implement?
```
Comment (thecardkid) [inline day07/README.md]: Fixed

File: day07/README.md (hunk 5 above)
Comment (thecardkid) [inline day07/README.md]: Clarified. It's a search phrase like "java AND programming"

File: day07/README.md (hunk 6 above)
Comment (thecardkid) [inline day07/README.md]: We'll see

**Pull request merged**